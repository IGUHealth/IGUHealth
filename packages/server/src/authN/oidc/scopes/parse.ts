import * as r4Sets from "@iguhealth/fhir-types/r4/sets";
import * as r4bSets from "@iguhealth/fhir-types/r4b/sets";

import { OIDCError } from "../middleware/oauth_error_handling.js";

export type OIDCScope = {
  type: "openid" | "profile" | "email" | "offline_access" | "online_access";
};

export type LaunchScope = {
  type: "launch";
};

export type LaunchTypeScope = {
  type: "launch-type";
  launchType: "encounter" | "patient";
};

export type SMARTResourceScope = {
  type: "smart-resource";
  level: "user" | "system" | "patient";
  scope: "resource" | "all";
  resourceType?: string;
  permissions: {
    create: boolean;
    read: boolean;
    update: boolean;
    delete: boolean;
    search: boolean;
  };
};

type SMARTScope =
  | SMARTResourceScope
  | LaunchTypeScope
  | LaunchScope
  | {
      type: "fhirUser";
    };

export type Scope = OIDCScope | SMARTScope;

/**
 * Validates a resource type from scope.
 * @param type some chunk from scope
 * @returns boolean if the resource type is valid.
 */
function validateResourceType(type: string): boolean {
  return r4Sets.resourceTypes.has(type) || r4bSets.resourceTypes.has(type);
}

function parsePermissions(methods: string): {
  create: boolean;
  read: boolean;
  update: boolean;
  delete: boolean;
  search: boolean;
} {
  if (methods === "*") {
    return {
      create: true,
      read: true,
      update: true,
      delete: true,
      search: true,
    };
  }

  const methodsObj = {
    create: false,
    read: false,
    update: false,
    delete: false,
    search: false,
  };
  for (const method of methods.split("")) {
    switch (method) {
      /**
       * Type level create
       */
      case "c": {
        methodsObj.create = true;
        break;
      }
      /**
       * Instance level read
       * Instance level vread
       * Instance level history
       */
      case "r": {
        methodsObj.read = true;
        break;
      }
      /**
       * Instance level update Note that some servers allow for an update operation to create a new instance,
       * and this is allowed by the update scope
       * Instance level patch
       */
      case "u": {
        methodsObj.update = true;
        break;
      }
      /**
       * Instance level delete
       */
      case "d": {
        methodsObj.delete = true;
        break;
      }
      /**
       * Type level search
       * Type level history
       * System level search
       * System level history
       */
      case "s": {
        methodsObj.search = true;
        break;
      }
      default: {
        throw new OIDCError({
          error: "invalid_scope",
          error_description: `Invalid scope access type methods: '${methods}' not supported.`,
        });
      }
    }
  }

  return methodsObj;
}

function isOIDCScope(
  scope: string,
): scope is
  | "openid"
  | "profile"
  | "email"
  | "offline_access"
  | "online_access" {
  return [
    "openid",
    "profile",
    "email",
    "offline_access",
    "online_access",
  ].includes(scope);
}

function validateSmartResourceLevel(
  level: string,
): level is "user" | "system" | "patient" {
  return ["user", "system", "patient"].includes(level);
}

export function parseScopes(scopes: string): Scope[] {
  return scopes
    .split(/\s/)
    .filter((scope) => scope !== "")
    .map((scope): Scope => {
      switch (true) {
        case isOIDCScope(scope): {
          return { type: scope };
        }
        case scope === "fhirUser": {
          return { type: scope };
        }
        case scope === "launch": {
          return { type: "launch" };
        }
        case scope.startsWith("launch"): {
          const chunks = scope.split("/");
          switch (chunks[1]) {
            case "patient":
            case "encounter": {
              return { type: "launch-type", launchType: chunks[1] };
            }
            default: {
              throw new OIDCError({
                error: "invalid_scope",
                error_description: `Invalid scope: '${scope}'.`,
              });
            }
          }
        }
        case scope.startsWith("user/"):
        case scope.startsWith("system/"):
        case scope.startsWith("patient/"): {
          const [level, permissionString] = scope.split("/");
          if (!permissionString) {
            throw new OIDCError({
              error: "invalid_scope",
              error_description: `Invalid scope: '${scope}'.`,
            });
          }
          if (!validateSmartResourceLevel(level)) {
            throw new OIDCError({
              error: "invalid_scope",
              error_description: `Invalid scope: '${scope}'.`,
            });
          }

          const permissions = permissionString.split(".");
          if (permissions.length !== 2) {
            throw new OIDCError({
              error: "invalid_scope",
              error_description: `Invalid scope: '${scope}'.`,
            });
          }

          const [resourceType, methods] = permissions;

          switch (true) {
            case resourceType === "*": {
              return {
                type: "smart-resource",
                level,
                scope: "all",
                permissions: parsePermissions(methods),
              };
            }
            case validateResourceType(resourceType): {
              return {
                type: "smart-resource",
                level,
                scope: "resource",
                resourceType,
                permissions: parsePermissions(methods),
              };
            }
            default: {
              throw new OIDCError({
                error: "invalid_scope",
                error_description: `Invalid scope: '${scope}'.`,
              });
            }
          }
        }
        default: {
          throw new OIDCError({
            error: "invalid_scope",
            error_description: `Invalid scope: '${scope}' not supported.`,
          });
        }
      }
    });
}

export function toString(scopes: Scope[]): string {
  return scopes
    .reduce((scopeString, scope: Scope) => {
      switch (scope.type) {
        case "online_access":
        case "offline_access":
        case "openid":
        case "profile":
        case "email": {
          return `${scopeString} ${scope.type}`;
        }
        case "fhirUser": {
          return `${scopeString} fhirUser`;
        }
        case "launch-type": {
          return `${scopeString} launch/${scope.launchType}`;
        }
        case "launch": {
          return `${scopeString} launch`;
        }
        case "smart-resource": {
          return `${scopeString} ${scope.level}/${scope.scope === "all" ? "*" : scope.resourceType}.${scope.permissions.create ? "c" : ""}${scope.permissions.read ? "r" : ""}${scope.permissions.update ? "u" : ""}${scope.permissions.delete ? "d" : ""}${scope.permissions.search ? "s" : ""}`;
        }
      }
    }, "")
    .slice(1);
}
