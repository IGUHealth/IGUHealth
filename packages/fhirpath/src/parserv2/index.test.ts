import { expect, test } from "@jest/globals";

import parse from "./index.js";

test("parserv2 primitive test", () => {
  expect(parse("45.5 + 2 * 5 = 27")).toMatchSnapshot();
  expect(parse("'hello'+'wolrd'")).toMatchSnapshot();
  expect(parse("@1980 + @T12:00")).toMatchSnapshot();
  expect(parse("@1980T12:00Z + @T12:00")).toMatchSnapshot();
});

test("Nested access ops", () => {
  expect(parse("$this.test != 45")).toMatchSnapshot();
});

test("Nested access", () => {
  expect(parse("$this.test")).toMatchSnapshot();
});

test("variables", () => {
  expect(parse("%hello.test + %myint_test")).toMatchSnapshot();
});

test("functions", () => {
  expect(parse("hello()")).toMatchSnapshot();
  expect(parse("hello(45)")).toMatchSnapshot();
  expect(parse("hello(45, 32)")).toMatchSnapshot();
  expect(parse("%deep.test(45,32)")).toMatchSnapshot();
});

test("paren", () => {
  expect(parse("($this.test)[$total]")).toMatchSnapshot();
});
