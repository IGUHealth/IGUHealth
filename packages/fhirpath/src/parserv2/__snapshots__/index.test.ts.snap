// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nested access 1`] = `
{
  "expression": [
    {
      "type": "this",
    },
    {
      "type": "identifier",
      "value": "test",
    },
  ],
  "type": "expression",
}
`;

exports[`Nested access ops 1`] = `
{
  "left": {
    "expression": [
      {
        "type": "this",
      },
      {
        "type": "identifier",
        "value": "test",
      },
    ],
    "type": "expression",
  },
  "operator": "!=",
  "right": {
    "expression": [
      {
        "literalType": "number",
        "type": "literal",
        "value": 45,
      },
    ],
    "type": "expression",
  },
  "type": "operation",
}
`;

exports[`functions 1`] = `
{
  "expression": [
    {
      "functionName": "hello",
      "parameters": [],
      "type": "function",
    },
  ],
  "type": "expression",
}
`;

exports[`functions 2`] = `
{
  "expression": [
    {
      "functionName": "hello",
      "parameters": [
        {
          "expression": [
            {
              "literalType": "number",
              "type": "literal",
              "value": 45,
            },
          ],
          "type": "expression",
        },
      ],
      "type": "function",
    },
  ],
  "type": "expression",
}
`;

exports[`functions 3`] = `
{
  "expression": [
    {
      "functionName": "hello",
      "parameters": [
        {
          "expression": [
            {
              "literalType": "number",
              "type": "literal",
              "value": 45,
            },
          ],
          "type": "expression",
        },
        {
          "expression": [
            {
              "literalType": "number",
              "type": "literal",
              "value": 32,
            },
          ],
          "type": "expression",
        },
      ],
      "type": "function",
    },
  ],
  "type": "expression",
}
`;

exports[`functions 4`] = `
{
  "expression": [
    {
      "type": "variable",
      "value": "deep",
    },
    {
      "functionName": "test",
      "parameters": [
        {
          "expression": [
            {
              "literalType": "number",
              "type": "literal",
              "value": 45,
            },
          ],
          "type": "expression",
        },
        {
          "expression": [
            {
              "literalType": "number",
              "type": "literal",
              "value": 32,
            },
          ],
          "type": "expression",
        },
      ],
      "type": "function",
    },
  ],
  "type": "expression",
}
`;

exports[`paren 1`] = `
{
  "expression": [
    {
      "expression": [
        {
          "type": "this",
        },
        {
          "type": "identifier",
          "value": "test",
        },
      ],
      "type": "expression",
    },
    {
      "index": {
        "expression": [
          {
            "type": "total",
          },
        ],
        "type": "expression",
      },
      "type": "indexed",
    },
  ],
  "type": "expression",
}
`;

exports[`parserv2 primitive test 1`] = `
{
  "left": {
    "left": {
      "expression": [
        {
          "literalType": "number",
          "type": "literal",
          "value": 45.5,
        },
      ],
      "type": "expression",
    },
    "operator": "+",
    "right": {
      "left": {
        "expression": [
          {
            "literalType": "number",
            "type": "literal",
            "value": 2,
          },
        ],
        "type": "expression",
      },
      "operator": "*",
      "right": {
        "expression": [
          {
            "literalType": "number",
            "type": "literal",
            "value": 5,
          },
        ],
        "type": "expression",
      },
      "type": "operation",
    },
    "type": "operation",
  },
  "operator": "=",
  "right": {
    "expression": [
      {
        "literalType": "number",
        "type": "literal",
        "value": 27,
      },
    ],
    "type": "expression",
  },
  "type": "operation",
}
`;

exports[`parserv2 primitive test 2`] = `
{
  "left": {
    "expression": [
      {
        "literalType": "string",
        "type": "literal",
        "value": "hello",
      },
    ],
    "type": "expression",
  },
  "operator": "+",
  "right": {
    "expression": [
      {
        "literalType": "string",
        "type": "literal",
        "value": "wolrd",
      },
    ],
    "type": "expression",
  },
  "type": "operation",
}
`;

exports[`parserv2 primitive test 3`] = `
{
  "left": {
    "expression": [
      {
        "literalType": "datetime",
        "type": "literal",
        "value": "1980",
      },
    ],
    "type": "expression",
  },
  "operator": "+",
  "right": {
    "expression": [
      {
        "literalType": "time",
        "type": "literal",
        "value": "12:00",
      },
    ],
    "type": "expression",
  },
  "type": "operation",
}
`;

exports[`parserv2 primitive test 4`] = `
{
  "left": {
    "expression": [
      {
        "literalType": "datetime",
        "type": "literal",
        "value": "1980T12:00Z",
      },
    ],
    "type": "expression",
  },
  "operator": "+",
  "right": {
    "expression": [
      {
        "literalType": "time",
        "type": "literal",
        "value": "12:00",
      },
    ],
    "type": "expression",
  },
  "type": "operation",
}
`;

exports[`variables 1`] = `
{
  "left": {
    "expression": [
      {
        "type": "variable",
        "value": "hello",
      },
      {
        "type": "identifier",
        "value": "test",
      },
    ],
    "type": "expression",
  },
  "operator": "+",
  "right": {
    "expression": [
      {
        "type": "variable",
        "value": "myint_test",
      },
    ],
    "type": "expression",
  },
  "type": "operation",
}
`;
